

<!DOCTYPE html PUBLIC "-//W3C//DTD XHTML 1.0 Transitional//EN"
  "http://www.w3.org/TR/xhtml1/DTD/xhtml1-transitional.dtd">

<html xmlns="http://www.w3.org/1999/xhtml">
  <head>
    <meta http-equiv="Content-Type" content="text/html; charset=utf-8" />
    
    <title>rnaseqlib &mdash; rnaseqlib v0.1 documentation</title>
    <link rel="stylesheet" href="_static/default.css" type="text/css" />
    <link rel="stylesheet" href="_static/pygments.css" type="text/css" />
    <script type="text/javascript">
      var DOCUMENTATION_OPTIONS = {
        URL_ROOT:    '',
        VERSION:     '0.1',
        COLLAPSE_INDEX: false,
        FILE_SUFFIX: '.html',
        HAS_SOURCE:  true
      };
    </script>
    <script type="text/javascript" src="_static/jquery.js"></script>
    <script type="text/javascript" src="_static/underscore.js"></script>
    <script type="text/javascript" src="_static/doctools.js"></script>
    <link rel="top" title="rnaseqlib v0.1 documentation" href="index.html" /> 
  </head>
  <body>
    <div class="related">
      <h3>Navigation</h3>
      <ul>
        <li class="right" style="margin-right: 10px">
          <a href="genindex.html" title="General Index"
             accesskey="I">index</a></li>
        <li><a href="#">rnaseqlib v0.1 documentation</a> &raquo;</li> 
      </ul>
    </div>  

    <div class="document">
      <div class="documentwrapper">
        <div class="bodywrapper">
          <div class="body">
            
  <div class="contents topic" id="contents">
<p class="topic-title first">Contents</p>
<ul class="simple">
<li><a class="reference internal" href="#rnaseqlib" id="id1"><tt class="docutils literal"><span class="pre">rnaseqlib</span></tt></a></li>
<li><a class="reference internal" href="#features" id="id2">Features</a></li>
<li><a class="reference internal" href="#updates" id="id3">Updates</a></li>
<li><a class="reference internal" href="#installation" id="id4">Installation</a></li>
<li><a class="reference internal" href="#running-rnaseqlib" id="id5">Running <tt class="docutils literal"><span class="pre">rnaseqlib</span></tt></a><ul>
<li><a class="reference internal" href="#initializing-an-rna-base-for-your-genome" id="id6">Initializing an RNA base for your genome</a></li>
<li><a class="reference internal" href="#configuration-specifying-your-data-and-mapping-parameters" id="id7">Configuration: specifying your data and mapping parameters</a></li>
<li><a class="reference internal" href="#running-the-pipeline" id="id8">Running the pipeline</a></li>
<li><a class="reference internal" href="#command-line-options" id="id9">Command-line options</a></li>
</ul>
</li>
<li><a class="reference internal" href="#frequently-asked-questions-faq" id="id10">Frequently Asked Questions (FAQ)</a></li>
<li><a class="reference internal" href="#authors" id="id11">Authors</a></li>
<li><a class="reference internal" href="#acknowledgements" id="id12">Acknowledgements</a></li>
</ul>
</div>
<div class="section" id="rnaseqlib">
<h1><a class="toc-backref" href="#id1"><tt class="docutils literal"><span class="pre">rnaseqlib</span></tt></a><a class="headerlink" href="#rnaseqlib" title="Permalink to this headline">¶</a></h1>
<p><tt class="docutils literal"><span class="pre">rnaseqlib</span></tt> is a simple, lightweight pipeline for RNA-Seq analysis.</p>
</div>
<div class="section" id="features">
<h1><a class="toc-backref" href="#id2">Features</a><a class="headerlink" href="#features" title="Permalink to this headline">¶</a></h1>
<ul class="simple">
<li>Pipeline for analyzing RNA sequencing data:<ul>
<li>Maps reads to genome and splice junctions</li>
<li>Computes basic quality control statistics</li>
<li>Outputs RPKM values for genes</li>
</ul>
</li>
<li>Supports several sequencing experiments:<ul>
<li>mRNA sequencing (mRNA-Seq)</li>
<li>Ribosome profiling data (Ribo-Seq)</li>
<li><em>CLIP for RNA-binding proteins (CLIP-Seq) [In progress]</em></li>
<li><em>SELEX-Seq [In progress]</em></li>
</ul>
</li>
<li>Contains utilities for processing <a class="reference external" href="http://genes.mit.edu/burgelab/miso/">MISO</a> output</li>
</ul>
</div>
<div class="section" id="updates">
<h1><a class="toc-backref" href="#id3">Updates</a><a class="headerlink" href="#updates" title="Permalink to this headline">¶</a></h1>
<p><strong>2012</strong></p>
<ul class="simple">
<li>Released <tt class="docutils literal"><span class="pre">rnaseqlib</span></tt></li>
</ul>
</div>
<div class="section" id="installation">
<h1><a class="toc-backref" href="#id4">Installation</a><a class="headerlink" href="#installation" title="Permalink to this headline">¶</a></h1>
<p>Get <tt class="docutils literal"><span class="pre">rnaseqlib</span></tt> from the GitHub repository (<a class="reference external" href="http://github.com/yarden/rnaseqlib">http://github.com/yarden/rnaseqlib</a>). To install with <cite>distribute</cite>&#8216;s <cite>easy_install</cite>:</p>
<div class="highlight-python"><pre>easy_install rnaseqlib</pre>
</div>
<p>For local installation, use:</p>
<div class="highlight-python"><pre>easy_install --user -U rnaseqlib</pre>
</div>
<p>If you don&#8217;t have <cite>easy_install</cite>, install it as an ordinary Python package:</p>
<div class="highlight-python"><pre>python setup.py install</pre>
</div>
<p>Or for local installation with distribute:</p>
<div class="highlight-python"><pre>python setup.py install --prefix=/your/local/dir</pre>
</div>
</div>
<div class="section" id="running-rnaseqlib">
<h1><a class="toc-backref" href="#id5">Running <tt class="docutils literal"><span class="pre">rnaseqlib</span></tt></a><a class="headerlink" href="#running-rnaseqlib" title="Permalink to this headline">¶</a></h1>
<div class="section" id="initializing-an-rna-base-for-your-genome">
<h2><a class="toc-backref" href="#id6">Initializing an RNA base for your genome</a><a class="headerlink" href="#initializing-an-rna-base-for-your-genome" title="Permalink to this headline">¶</a></h2>
<p>The first step is to create a set of files, called an <em>RNA Base</em>, required to run <tt class="docutils literal"><span class="pre">rnaseqlib</span></tt> for a particular genome. For example,</p>
<div class="highlight-python"><pre>rna_pipeline.py --init mm9 --output-dir pipeline_init</pre>
</div>
<p>This will create a directory called <tt class="docutils literal"><span class="pre">mm9</span></tt> in <tt class="docutils literal"><span class="pre">pipeline_init</span></tt> containing the necessary files for running the pipeline on the mm9 genome.
The initialization procedure will, among other things, do the following:</p>
<blockquote>
<div><ul class="simple">
<li>Download the genome sequence files from UCSC (in FASTA format)</li>
<li>Download gene tables from UCSC for Ensembl genes, UCSC knownGenes, and RefSeq genes.</li>
<li>Computate the coordinates for exons, introns, constitutive exons, constitutive exons in coding regions, and other
useful features of gene tables</li>
<li>Index the genome files using <tt class="docutils literal"><span class="pre">bowtie-build</span></tt></li>
</ul>
</div></blockquote>
</div>
<div class="section" id="configuration-specifying-your-data-and-mapping-parameters">
<h2><a class="toc-backref" href="#id7">Configuration: specifying your data and mapping parameters</a><a class="headerlink" href="#configuration-specifying-your-data-and-mapping-parameters" title="Permalink to this headline">¶</a></h2>
<p id="config">The settings of the RNA-Seq pipeline are specified through a single settings file that contains four main sections:</p>
<blockquote>
<div><ul class="simple">
<li><tt class="docutils literal"><span class="pre">pipeline</span></tt>: what data type is used (e.g. RNA-Seq, Ribo-Seq, etc.)</li>
<li><tt class="docutils literal"><span class="pre">pipeline-files</span></tt>: where the pipeline initialization files are stored</li>
<li><tt class="docutils literal"><span class="pre">mapping</span></tt>: parameters related to mapping of the data (e.g. what mapper to use, where the genome index is.)</li>
<li><tt class="docutils literal"><span class="pre">data</span></tt>: where the input sequence files are and where the results should be outputted</li>
</ul>
</div></blockquote>
<p>The following is an example settings file for a set of mRNA-Seq samples:</p>
<div class="highlight-python"><pre>##
## Pipeline settings for RNA-Seq
##
[pipeline]
# Data type to use (e.g. 'rnaseq' or 'riboseq')
data_type = rnaseq

[pipeline-files]
init_dir = ./mm9

[mapping]
mapper = tophat
bowtie_path = bowtie
tophat_path = tophat
tophat_index = mm9_index
cluster_type = bsub
genome = mm9
tophat_options = --bowtie1 --min-anchor-length 4
tophat_gtf = ./mm9/ucsc/knownGene.gtf
readlen = 40
overhanglen = 4
paired = True
paired_end_frag = 300
stranded = fr-first

[data]
indir = ./input_dir
outdir = ./results

sequence_files = [
    ["sample1_p1.fastq.gz", "sample1_p1"],
    ["sample1_p2.fastq.gz", "sample1_p2"],
    ["sample2_p1.fastq.gz", "sample2_p1"],
    ["sample2_p2.fastq.gz", "sample2_p2"]]

sample_groups = [["sample1", ["sample1_p1", "sample1_p2"]],
                 ["sample2", ["sample2_p1", "sample2_p2"]]]</pre>
</div>
</div>
<div class="section" id="running-the-pipeline">
<h2><a class="toc-backref" href="#id8">Running the pipeline</a><a class="headerlink" href="#running-the-pipeline" title="Permalink to this headline">¶</a></h2>
<p>To run the pipeline, use the <tt class="docutils literal"><span class="pre">--run</span></tt> option:</p>
<div class="highlight-python"><pre>rna_pipeline.py --run --settings ./settings.txt --output-dir ./my_results</pre>
</div>
<p>where <tt class="docutils literal"><span class="pre">settings.txt</span></tt> is the pipeline settings file and <tt class="docutils literal"><span class="pre">my_results</span></tt> is a directory where the pipeline output should go.</p>
</div>
<div class="section" id="command-line-options">
<h2><a class="toc-backref" href="#id9">Command-line options</a><a class="headerlink" href="#command-line-options" title="Permalink to this headline">¶</a></h2>
<p><tt class="docutils literal"><span class="pre">run_pipeline.py</span></tt> is the main driver script of the pipeline. It takes the following main arguments:</p>
<div class="highlight-python"><pre>--run                 Run pipeline.

--settings
                      Settings filename.

--init
                      Initialize the pipeline. Takes as input a genome, e.g.
                      mm9 or hg18.

--output-dir
                      Output directory.</pre>
</div>
</div>
</div>
<div class="section" id="frequently-asked-questions-faq">
<h1><a class="toc-backref" href="#id10">Frequently Asked Questions (FAQ)</a><a class="headerlink" href="#frequently-asked-questions-faq" title="Permalink to this headline">¶</a></h1>
<div class="admonition note">
<p class="first admonition-title">Note</p>
<p class="last">Section under construction</p>
</div>
</div>
<div class="section" id="authors">
<span id="katz"></span><span id="refs"></span><h1><a class="toc-backref" href="#id11">Authors</a><a class="headerlink" href="#authors" title="Permalink to this headline">¶</a></h1>
<p><tt class="docutils literal"><span class="pre">rnaseqlib</span></tt> was written by Yarden Katz.</p>
</div>
<div class="section" id="acknowledgements">
<h1><a class="toc-backref" href="#id12">Acknowledgements</a><a class="headerlink" href="#acknowledgements" title="Permalink to this headline">¶</a></h1>
<p>Thanks to:</p>
<ul class="simple">
<li>Noah Spies (MIT)</li>
<li>Eric Wang (MIT)</li>
</ul>
</div>


          </div>
        </div>
      </div>
      <div class="sphinxsidebar">
        <div class="sphinxsidebarwrapper">
  <h3><a href="#">Table Of Contents</a></h3>
  <ul>
<li><a class="reference internal" href="#"><tt class="docutils literal"><span class="pre">rnaseqlib</span></tt></a></li>
<li><a class="reference internal" href="#features">Features</a></li>
<li><a class="reference internal" href="#updates">Updates</a></li>
<li><a class="reference internal" href="#installation">Installation</a></li>
<li><a class="reference internal" href="#running-rnaseqlib">Running <tt class="docutils literal"><span class="pre">rnaseqlib</span></tt></a><ul>
<li><a class="reference internal" href="#initializing-an-rna-base-for-your-genome">Initializing an RNA base for your genome</a></li>
<li><a class="reference internal" href="#configuration-specifying-your-data-and-mapping-parameters">Configuration: specifying your data and mapping parameters</a></li>
<li><a class="reference internal" href="#running-the-pipeline">Running the pipeline</a></li>
<li><a class="reference internal" href="#command-line-options">Command-line options</a></li>
</ul>
</li>
<li><a class="reference internal" href="#frequently-asked-questions-faq">Frequently Asked Questions (FAQ)</a></li>
<li><a class="reference internal" href="#authors">Authors</a></li>
<li><a class="reference internal" href="#acknowledgements">Acknowledgements</a></li>
</ul>

  <h3>This Page</h3>
  <ul class="this-page-menu">
    <li><a href="_sources/rnaseqlib.txt"
           rel="nofollow">Show Source</a></li>
  </ul>
<div id="searchbox" style="display: none">
  <h3>Quick search</h3>
    <form class="search" action="search.html" method="get">
      <input type="text" name="q" size="18" />
      <input type="submit" value="Go" />
      <input type="hidden" name="check_keywords" value="yes" />
      <input type="hidden" name="area" value="default" />
    </form>
    <p class="searchtip" style="font-size: 90%">
    Enter search terms or a module, class or function name.
    </p>
</div>
<script type="text/javascript">$('#searchbox').show(0);</script>
        </div>
      </div>
      <div class="clearer"></div>
    </div>
    <div class="related">
      <h3>Navigation</h3>
      <ul>
        <li class="right" style="margin-right: 10px">
          <a href="genindex.html" title="General Index"
             >index</a></li>
        <li><a href="#">rnaseqlib v0.1 documentation</a> &raquo;</li> 
      </ul>
    </div>
    <div class="footer">
        &copy; Copyright 2012, Yarden Katz.
      Created using <a href="http://sphinx.pocoo.org/">Sphinx</a> 1.0.7.
    </div>
  </body>
</html>